libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -362,6 +365,77 @@ struct NoVariablesInInterfaceChecker: public PostTypeChecker::Checker
 	/// Flag indicating whether we are currently inside a StructDefinition.
 	int m_insideStruct = 0;
 };
+
+struct CircularContractDependencies: public PostTypeChecker::Checker
+{
+	CircularContractDependencies(ErrorReporter& _errorReporter):
+		Checker(_errorReporter)
+	{}
+
+	bool visit(ContractDefinition const& _contract) override
+	{
+		auto joined = boost::join(
+			_contract.annotation().creationCallGraph->get()->createdContracts,
+			_contract.annotation().deployedCallGraph->get()->createdContracts
+		);
+
+		for (auto const& [contract, referencee]: joined)
+			if (contract != &_contract)
+				_contract.annotation().contractDependencies.emplace(contract, referencee);
+
+		return true;
+	}
+
+	void endVisit(ContractDefinition const& _contract) override
+	{
+		auto const& [contract, referencee] = dependenciesAreCyclic(_contract);
+
+		if (contract)
+			m_foundCycles.emplace(contract, referencee);
+	}
+
+	void finalize() override
+	{
+		for (auto const& [cycle, referencee]: m_foundCycles)
+		{
+			SecondarySourceLocation ssl{};
+			ssl.append(string{"Referenced contract"}, cycle->location());
+
+			m_errorReporter.typeError(
+				7813_error,
+				referencee->location(),
+				ssl,
```

Comment:
I see that `SecondarySourceLocation` is just a plain struct so this should work:

```suggestion
				{{"Referenced contract"s, cycle->location()}},
```

Reply:
Aah, I was wondering how to do that best.

Reply:
Hmm unfortunately that doesn't work

Reply:
Oh. Weird. Maybe it requires the name of the type to be given explicitly?

Anyway, I guess it's just a small detail not worth digging into too much so I'll just mark it as resolved :)